File: jacsnpm/verification.ts
/**
 * Fail-closed helpers for presenting verification results.
 *
 * Legacy-v1 compatibility authenticates payload fields, not signer/key/time
 * metadata. Centralizing this rule prevents high-level APIs from accidentally
 * presenting parsed legacy fields as authenticated attribution.
 */

export const V2_SIGNATURE_CONTENT_VERSION = 'jacs-signature-v2';

/**
 * Require an exact successful result from a native cryptographic verifier.
 * High-level wrappers must never infer success from completion, truthiness, or
 * a malformed binding return value.
 */
export function requireLiteralTrueVerification(
  result: unknown,
  context: string,
): void {
  if (result !== true) {
    throw new Error(`${context} did not return literal true`);
  }
}

export interface AuthenticatedSignatureMetadata {
  signerId: string;
  publicKeyHash: string;
  timestamp: string;
}

export function authenticatedSignatureMetadata(
  document: unknown,
): AuthenticatedSignatureMetadata {
  if (!document || typeof document !== 'object' || Array.isArray(document)) {
    return { signerId: '', publicKeyHash: '', timestamp: '' };
  }
  const signature = (document as Record<string, unknown>).jacsSignature;
  if (!signature || typeof signature !== 'object' || Array.isArray(signature)) {
    return { signerId: '', publicKeyHash: '', timestamp: '' };
  }
  const metadata = signature as Record<string, unknown>;
  if (metadata.signatureContentVersion !== V2_SIGNATURE_CONTENT_VERSION) {
    return { signerId: '', publicKeyHash: '', timestamp: '' };
  }

  const signerId = metadata.agentId ?? metadata.agentID;
  return {
    signerId: typeof signerId === 'string' ? signerId : '',
    publicKeyHash: typeof metadata.publicKeyHash === 'string' ? metadata.publicKeyHash : '',
    timestamp: typeof metadata.date === 'string' ? metadata.date : '',
  };
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export interface NormalizedAgreementStatus {
  complete: boolean;
  signers: Array<{
    agentId: string;
    signed: boolean;
    signedAt?: string;
  }>;
  pending: string[];
}

/** Normalize agreement status without allowing JSON truthiness to grant completion. */
export function normalizeAgreementStatus(result: unknown): NormalizedAgreementStatus {
  if (!isRecord(result)) {
    throw new Error('Agreement status result must be a JSON object');
  }
  const rawSigners = result.signers;
  const signersWellFormed = Array.isArray(rawSigners)
    && rawSigners.every((signer) => isRecord(signer) && typeof signer.signed === 'boolean');
  const signers = Array.isArray(rawSigners)
    ? rawSigners.map((value) => {
      const signer = isRecord(value) ? value : {};
      const rawAgentId = signer.agentId ?? signer.agent_id;
      const rawSignedAt = signer.signedAt ?? signer.signed_at;
      return {
        agentId: typeof rawAgentId === 'string' ? rawAgentId : '',
        signed: signer.signed === true,
        ...(typeof rawSignedAt === 'string' ? { signedAt: rawSignedAt } : {}),
      };
    })
    : [];
  const pendingWellFormed = Array.isArray(result.pending)
    && result.pending.every((agentId) => typeof agentId === 'string');
  const pending = pendingWellFormed ? result.pending as string[] : [];
  return {
    complete: result.complete === true && signersWellFormed && pendingWellFormed,
    signers,
    pending,
  };
}

/**
 * Normalize an attestation result and recompute its security aggregate from
 * exact booleans. Missing, string, numeric, or contradictory flags fail closed.
 */
export function normalizeAttestationVerificationResult(
  result: unknown,
): Record<string, unknown> {
  if (!isRecord(result)) {
    throw new Error('Attestation verification result must be a JSON object');
  }
  const normalized: Record<string, unknown> = { ...result };

  const crypto = isRecord(result.crypto) ? { ...result.crypto } : {};
  const signatureValid = crypto.signatureValid === true;
  const hashValid = crypto.hashValid === true;
  crypto.signatureValid = signatureValid;
  crypto.hashValid = hashValid;
  normalized.crypto = crypto;
  const cryptoValid = signatureValid && hashValid;

  const rawEvidence = result.evidence;
  const evidenceWellFormed = Array.isArray(rawEvidence);
  const evidence = evidenceWellFormed
    ? rawEvidence.map((value: unknown) => {
      const item = isRecord(value) ? { ...value } : {};
      item.digestValid = item.digestValid === true;
      item.freshnessValid = item.freshnessValid === true;
      return item;
    })
    : [];
  normalized.evidence = evidence;
  const evidenceValid = evidenceWellFormed
    && evidence.every(
      (item: Record<string, unknown>) => (
        item.digestValid === true && item.freshnessValid === true
      ),
    );

  let chainValid = true;
  if (result.chain === null || result.chain === undefined) {
    normalized.chain = null;
  } else {
    const chain = isRecord(result.chain) ? { ...result.chain } : {};
    const rawLinks = chain.links;
    const linksWellFormed = Array.isArray(rawLinks);
    const links = linksWellFormed
      ? rawLinks.map((value: unknown) => {
        const link = isRecord(value) ? { ...value } : {};
        link.valid = link.valid === true;
        return link;
      })
      : [];
    chain.links = links;
    const normalizedChainValid = chain.valid === true
      && linksWellFormed
      && links.every((link: Record<string, unknown>) => link.valid === true);
    chain.valid = normalizedChainValid;
    normalized.chain = chain;
    chainValid = normalizedChainValid;
  }

  normalized.valid = result.valid === true
    && cryptoValid
    && evidenceValid
    && chainValid;
  return normalized;
}

File: jacspy/python/jacs/_verification.py
"""Fail-closed helpers for presenting verification results.

Cryptographic legacy-v1 compatibility authenticates document payload fields,
but not the signer, key, or time fields stored under ``jacsSignature``. Keep
that distinction in one place so every high-level Python API presents the same
attribution contract.
"""

from typing import Any, Mapping


V2_SIGNATURE_CONTENT_VERSION = "jacs-signature-v2"


def authenticated_signature_metadata(
    document: Mapping[str, Any] | None,
    *,
    verified: bool,
) -> tuple[str, str, str]:
    """Return attribution only for a literally valid v2 verification result.

    A v2-shaped envelope does not authenticate itself. Callers must supply the
    native verifier's result so invalid documents can never promote parsed
    signer, key, or timestamp fields into trusted high-level output.
    """
    if verified is not True:
        return "", "", ""
    if not isinstance(document, Mapping):
        return "", "", ""
    signature = document.get("jacsSignature")
    if not isinstance(signature, Mapping):
        return "", "", ""
    if signature.get("signatureContentVersion") != V2_SIGNATURE_CONTENT_VERSION:
        return "", "", ""

    signer_id = signature.get("agentId", signature.get("agentID", ""))
    public_key_hash = signature.get("publicKeyHash", "")
    timestamp = signature.get("date", "")
    return (
        signer_id if isinstance(signer_id, str) else "",
        public_key_hash if isinstance(public_key_hash, str) else "",
        timestamp if isinstance(timestamp, str) else "",
    )


def normalize_attestation_verification_result(result: Any) -> dict[str, Any]:
    """Normalize an attestation result without trusting JSON truthiness.

    The native result is an authenticated output contract. A missing or
    non-boolean security flag therefore cannot be treated as success. The
    aggregate result is additionally ANDed with every nested cryptographic,
    evidence, and chain flag so contradictory native output fails closed.
    """
    if not isinstance(result, Mapping):
        raise ValueError("attestation verification result must be a JSON object")

    normalized = dict(result)

    raw_crypto = result.get("crypto")
    crypto = dict(raw_crypto) if isinstance(raw_crypto, Mapping) else {}
    crypto["signatureValid"] = crypto.get("signatureValid") is True
    crypto["hashValid"] = crypto.get("hashValid") is True
    normalized["crypto"] = crypto
    crypto_valid = crypto["signatureValid"] and crypto["hashValid"]

    raw_evidence = result.get("evidence")
    evidence_well_formed = isinstance(raw_evidence, list)
    evidence = []
    if evidence_well_formed:
        for item in raw_evidence:
            entry = dict(item) if isinstance(item, Mapping) else {}
            entry["digestValid"] = entry.get("digestValid") is True
            entry["freshnessValid"] = entry.get("freshnessValid") is True
            evidence.append(entry)
    normalized["evidence"] = evidence
    evidence_valid = evidence_well_formed and all(
        item["digestValid"] and item["freshnessValid"] for item in evidence
    )

    raw_chain = result.get("chain")
    if raw_chain is None:
        normalized["chain"] = None
        chain_valid = True
    else:
        chain = dict(raw_chain) if isinstance(raw_chain, Mapping) else {}
        raw_links = chain.get("links")
        links_well_formed = isinstance(raw_links, list)
        links = []
        if links_well_formed:
            for item in raw_links:
                link = dict(item) if isinstance(item, Mapping) else {}
                link["valid"] = link.get("valid") is True
                links.append(link)
        chain["links"] = links
        chain["valid"] = (
            chain.get("valid") is True
            and links_well_formed
            and all(link["valid"] for link in links)
        )
        normalized["chain"] = chain
        chain_valid = chain["valid"]

    normalized["valid"] = (
        result.get("valid") is True and crypto_valid and evidence_valid and chain_valid
    )
    return normalized
